hack/autodoc/autodoc.go

```diff
@@ -0,0 +1,347 @@
+		os.Exit(errorExitCode)
+	}
+
+	constReasons := getConstReasons(dataChunks[0])
+
+	errors, reasonsMetadata := getAndValidateReasonsMetadata(dataChunks[1])
+	if len(errors) > 0 {
+		printErrors(errors)
+		os.Exit(errorExitCode)
+	}
+
+	errors = checkIfConstsAndMetadataAreInSync(constReasons, reasonsMetadata)
+	if len(errors) > 0 {
+		fmt.Println("The declared reasons in const Go section are out out sync with Reasons metadata")
+		printErrors(errors)
+		os.Exit(errorExitCode)
+	}
+
+	mdTableContent := mdTableToStruct(dataChunks[2])
+	errors = compareContent(mdTableContent, reasonsMetadata)
+	if len(errors) > 0 {
+		printErrors(errors)
+		fmt.Println("Below can be found auto-generated table which contain new changes:")
+		fmt.Println(buildMdTable(reasonsMetadata))
+		os.Exit(errorExitCode)
+	}
+
+	os.Exit(okExitCode)
+}
+
+func extractData() string {
+	cmd := exec.Command("/bin/sh", "hack/autodoc/extract_conditions_data.sh")
+	var cmdOut, cmdErr bytes.Buffer
+	cmd.Stdout = &cmdOut
+	cmd.Stderr = &cmdErr
+	if err := cmd.Run(); err != nil {
+		fmt.Println(cmdErr.String())
+		os.Exit(errorExitCode)
+	}
+	return cmdOut.String()
+}
+
+func getConstReasons(input string) []string {
+	constReasons := make([]string, 0)
+	lines := strings.Split(input, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if len(fields) > 0 {
+			constReasons = append(constReasons, fields[0])
+		}
+	}
+	return constReasons
+}
+
+func getAndValidateReasonsMetadata(input string) ([]string, []reasonMetadata) {
+	lines := strings.Split(input, "\n")
+	reasonsMetadata := make([]reasonMetadata, 0)
+	errors := make([]string, 0)
+	for _, line := range lines {
+		if line == "" {
+			continue
+		}
+		err, lineStructured := tryConvertGoLineToStruct(line)
+		if err != nil {
+			errors = append(errors, err.Error())
+			continue
+		}
+		if lineStructured != nil {
+			reasonsMetadata = append(reasonsMetadata, *lineStructured)
+		}
+	}
+
+	return errors, reasonsMetadata
+}
+
+func checkIfConstsAndMetadataAreInSync(constReasons []string, reasonsMetadata []reasonMetadata) []string {
+	checkIfConstReasonHaveMetadata := func(constReason string) bool {
+		for _, reasonMetadata := range reasonsMetadata {
+			if reasonMetadata.conditionReason == constReason {
+				return true
+			}
+		}
+		return false
+	}
+	errors := make([]string, 0)
+
+	for _, constReason := range constReasons {
+		if !checkIfConstReasonHaveMetadata(constReason) {
+			errors = append(errors, fmt.Sprintf("there is a Reason = (%s) declarated in const scope, but there is no matching metadata for it", constReason))
+		}
+	}
+	return errors
+}
+
+func mdTableToStruct(tableMd string) []reasonMetadata {
+	mdRows := strings.Split(tableMd, "\n")
+	mdRows = mdRows[2 : len(mdRows)-1]
+	structuredData := make([]reasonMetadata, 0)
+	for _, mdRow := range mdRows {
+		cleanLine := strings.Split(mdRow, "|")
+		cleanLine = cleanLine[1 : len(cleanLine)-1]
+
+		crState := cleanLine[1]
```

Comment:
Here (and below) we risk accessing non-existing elements (if MD table is corrupted).